### ISSUE TITLE
How to run two instances on the same machine?

### ISSUE THREAD

**Issue by TangentFoxy** [rr-/szurubooru#301]
Tl;dr: I have a docker network providing a reverse proxy, and am trying to set up a second instance on a different domain. Everything worked for the first instance, but the second one on page load says that it cannot retrieve configuration.

----

I have successfully set up szurubooru on my server, with almost no issues. I wish to set up a second instance on another domain on the same machine.

The first issue I ran into is that I would need to rename the entries in docker-compose, or else they wouldn't be unique on the network, no big deal as far as I can tell.

The second issue is with setting `PORT` in `.env`. I must have misunderstood something, because I thought I would not have to modify this because they are running as separate containers. I'm connecting everything through a docker network so nothing is exposed except through the reverse proxy. When I remove the `ports` section from the docker file, or set it to anything other than `8080`, connecting to the website *works* but I get an error about failing to retrieve configuration.

I've been playing with settings for a few days and cannot figure out how to get the second instance working, while the first has run fine this entire time.

**Comment by TiredSysOp** [comment]
A sure fire way would be to use nginx in front and then reverse proxy to the szurubooru front end, create a vhost for each instance and change the domain and proxypass port, use different ports in .env or you'll get port conflicts

an example vhost config

server {
  server_name ricardodb.tk;
  return 301 https://ricardodb.tk$request_uri$is_args$args;
}

server	{
        listen                  443 ssl;
        server_name             ricardodb.tk;

		client_max_body_size 100M;
		client_body_timeout 30s;
		server_tokens off;
		location / {
			proxy_http_version 1.1;
			proxy_pass http://127.0.0.1:8500;
			proxy_set_header Host $http_host;
			proxy_set_header Upgrade $http_upgrade;
			proxy_set_header Connection "upgrade";
			proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
			proxy_set_header X-Scheme $scheme;
			proxy_set_header X-Real-IP $remote_addr;
			proxy_set_header X-Forwarded-Proto $scheme;
			proxy_set_header X-Script-Name /szuru;
			error_page 500 501 502 504 505 506 507 508 509 510 511 @err;
		}

		location @err {
			return 500 "server error. please try again later.";
			default_type text/plain;
		}

        ssl_certificate         /etc/letsencrypt/live/ricardodb.tk/cert.pem;
        ssl_certificate_key     /etc/letsencrypt/live/ricardodb.tk/privkey.pem;
}

**Comment by sgsunder** [comment]
I'm not really sure of the details of what you're trying to do, but the easiest way to run two instances would be to create two separate directories, and keep separate copies of `docker-compose.yml` and `.env` in each directory. You can then configure `.env` individually for each. You would have to ensure that `MOUNT_DATA` and `MOUNT_SQL` are separate for each instance, and that each instance is configured for a different `PORT`.

> The second issue is with setting PORT in .env. I must have misunderstood something, because I thought I would not have to modify this because they are running as separate containers.

`PORT` here is the port on your host machine that gets mapped into your container. So if you configured your first instance with `PORT=8080` and your second machine with `PORT=8081`, then you would access your first instance at `http://<host_machine_IP>:8080/` and your second instance at `http://<host_machine_IP>:8081/`

**Comment by TangentFoxy** [comment]
@TiredSysOp They are different containers entirely, why would there be port conflicts on completely separate interfaces on completely separate containers?

@sgsunder That's not at all how that's working as far as I can tell, because the primary instance is running with that set to 8080 but is perfectly accessible and usable on port 80 externally. Setting it to anything else makes the website accessible on port 80 as usual, but the website is not able to access its own backend.

It's acting like something is hardcoded as far as I can tell, because no matter how I change the configuration, nothing works on a second instance. Only the original unmodified instance is working.

**Comment by TiredSysOp** [comment]
I might be asking something silly, could you post your configs? Do you need to use docker networking? When using szurubooru it creates it's own "network" and exposes everything through a single port:
```
# Port to expose HTTP service
# Set to 127.0.0.1:8080 if you wish to reverse-proxy the docker's port,
# otherwise the port specified here will be publicly accessible
PORT=8500
```
To make life easier for yourself do not modify any docker compose files, the only things that need to be modified or added are the .env and config.yaml files. In all honesty you only need the .env file for it to "just work"

The following need to be different in order for both instances to run as intended
PORT=8500 This needs to be changed to avoid port conflicts within the host operating system
MOUNT_DATA=/var/local/szurubooru/data You don't want the instances to share the same folder when it comes to storing data
MOUNT_SQL=/var/local/szurubooru/sql users and post data is stored here, I don't need to say why this should not be the same

Then sort the domain and reverse proxying through nginx

**Comment by TangentFoxy** [comment]
I am using docker networking, I have a reverse-proxy container acting as an interface between various containers so that I can host multiple things on the same port with different domains.

I have attempted this again with fresh eyes, and it is working. I don't know what has happened with this project, but I notice it is no longer using elastisearch and that the docker-config is a lot simpler. Only downside is that I built the images locally instead of just using the existing ones, since those are based on an older build.

This is my current config for the first instance:

````yaml
version: '3.5'

networks:
  web:
    external: true

services:

  server:
    image: szurubooru/server:latest
    depends_on:
      - sql
      - elasticsearch
    environment:
      ## These should be the names of the dependent containers listed above,
      ## or FQDNs/IP addresses if these services are running outside of Docker
      POSTGRES_HOST: sql
      ESEARCH_HOST: elasticsearch
      ## Credentials for database:
      POSTGRES_USER:
      POSTGRES_PASSWORD:
      ## Commented Values are Default:
      #POSTGRES_DB: defaults to same as POSTGRES_USER
      #POSTGRES_PORT: 5432
      #ESEARCH_PORT: 9200
      #ESEARCH_INDEX: szurubooru
      #ESEARCH_PASSWORD: (empty by default, set if you are using an external
      #                  source for elasticsearch and want to use HTTP basic
      #                  authentication for security)
      #ESEARCH_USER: szurubooru (only used if password is set)
      #LOG_SQL: 0 (1 for verbose SQL logs)
    volumes:
      - "${MOUNT_DATA}:/data"
      - "./server/config.yaml:/opt/app/config.yaml"
    #restart: unless-stopped

  client:
    image: szurubooru/client:latest
    depends_on:
      - server
    environment:
      BACKEND_HOST: server
      BASE_URL:
    volumes:
      - "${MOUNT_DATA}:/data:ro"
    ports:
      - "${PORT}:80"
    networks:
      - default
      - web
    #restart: unless-stopped

  sql:
    image: postgres:11-alpine
    #restart: unless-stopped
    environment:
      POSTGRES_USER:
      POSTGRES_PASSWORD:
    volumes:
      - "${MOUNT_SQL}:/var/lib/postgresql/data"

  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.3.1
    environment:
      ## Specifies the Java heap size used
      ## Read
      ##  https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html
      ## for more info
      ES_JAVA_OPTS: -Xms512m -Xmx512m
    volumes:
      - index:/usr/share/elasticsearch/data
    ulimits:
      nofile:
        soft: 65536
        hard: 65536
    #restart: unless-stopped

volumes:
  index: # Scratch space for ElasticSearch index, will be rebuilt if lost
````

And this is the config I'm using for the second instance:

````yaml
version: '3.5'

networks:
  web:
    external: true

services:

  server:
    #image: szurubooru/server:latest
    build: "./server"
    image: jayserver
    depends_on:
      - sql
    environment:
      ## These should be the names of the dependent containers listed below,
      ## or FQDNs/IP addresses if these services are running outside of Docker
      POSTGRES_HOST: sql
      ## Credentials for database:
      POSTGRES_USER:
      POSTGRES_PASSWORD:
      ## Commented Values are Default:
      #POSTGRES_DB: defaults to same as POSTGRES_USER
      #POSTGRES_PORT: 5432
      #LOG_SQL: 0 (1 for verbose SQL logs)
    volumes:
      - "${MOUNT_DATA}:/data"
      - "./server/config.yaml:/opt/app/config.yaml"

  aliceclient:
    #image: szurubooru/client:latest
    build: "./client"
    image: aliceclient
    depends_on:
      - server
    environment:
      BACKEND_HOST: server
      BASE_URL:
    volumes:
      - "${MOUNT_DATA}:/data:ro"
    ports:
      - "${PORT}:80"
    networks:
      - default
      - web

  sql:
    image: postgres:11-alpine
    restart: unless-stopped
    environment:
      POSTGRES_USER:
      POSTGRES_PASSWORD:
    volumes:
      - "${MOUNT_SQL}:/var/lib/postgresql/data"
````

Thank you for your help. ~~I don't know exactly what went wrong before but everything is working alright now.~~ I think the issue I had before was that I was previous attempting to expose two containers on the same network (`web`) with the same name. I had a couple final issues before sorting everything out that had to do with that.

**Comment by TangentFoxy** [comment]
I've updated the original server to the latest commits from the master branch, and now the second server has mysteriously quit working again. I'm a bit at a loss as to what went wrong now.

They are running on independent networks, with the clients both on a shared network that I use for reverse proxying multiple services, with different container names for the clients and servers (I briefly had them both connecting to the same backend server, and changing the names stopped this), and the only other difference is `PORT=127.0.0.1:8081` on the second server, because the first one is taking port 8080.

That configuration was working when they were on different versions and somehow has stopped. Does anyone have a clue what I'm doing wrong now?

**Comment by sgsunder** [comment]
1) The functionality that we previously relied on elasticsearch for has now been fully integrated into Postgres as of db0c33b, so unless you froze onto a commit prior to this, you can safely remove the `elasticsearch` container and the `index` volume without any loss of data or functionality.

2) Double check to make sure that `MOUNT_DATA` and `MOUNT_SQL` are different on both instances. If you have both instances trying to read and write the same SQL tables, you are gonna have a bad time.

3) As far as I can tell, you can't specify an IP or CIDR for `PORT` (see [link](https://docs.docker.com/compose/compose-file/#ports)), so you might have to change `PORT=127.0.0.1:8001` to `PORT=8001`, and then use a firewall on your host machine instead.

**Comment by TangentFoxy** [comment]
1. This was taken care of.

2. They are. (Technically the same, but specified as relative paths, in completely different directories.)

3. ~~I'll give this a try right now.~~ I meant to do this a while ago, and forgot. Coming back to find I had left this comment open and unfinished.. well, it's somehow working again. I don't understand why, but *shrugs*.
